src/process_utility.c

```diff
@@ -1210,96 +1275,288 @@ process_index_chunk(Hypertable *ht, Oid chunk_relid, void *arg)
 }
 
 static void
-process_index_start(Node *parsetree)
+process_index_chunk_multitransaction(Hypertable *ht, Oid chunk_relid, void *arg)
 {
-	IndexStmt *stmt = (IndexStmt *) parsetree;
-	Cache *hcache;
-	Hypertable *ht;
+	CreateIndexInfo *info = (CreateIndexInfo *) arg;
+	CatalogSecurityContext sec_ctx;
+	Chunk *chunk;
+	Relation hypertable_rel;
+	Relation hypertable_index_rel;
+	Relation chunk_rel;
 
-	Assert(IsA(stmt, IndexStmt));
+	Assert(info->extended_options.multitransaction);
 
 	/*
-	 * PG11 adds some cases where the relation is not there, namely on
-	 * declaratively partitioned tables, with partitioned indexes:
-	 * https://github.com/postgres/postgres/commit/8b08f7d4820fd7a8ef6152a9dd8c6e3cb01e5f99
-	 * we don't deal with them so we will just return immediately
+	 * commit the previous transaction (foreach_chunk requires a transcation
+	 * running, so we cannot commit before this point)
 	 */
-	if (NULL == stmt->relation)
+	CommitTransactionCommand();
+
+	/* Start a new transaction for each relation. */
+	StartTransactionCommand();
+	MemoryContextSwitchTo(info->mctx);
+
+#ifdef DEBUG
+	if (info->extended_options.max_chunks == 0)
 		return;
 
-	hcache = ts_hypertable_cache_pin();
-	ht = ts_hypertable_cache_get_entry_rv(hcache, stmt->relation);
+	/*
+	 * if max_chunks is < 0 then we're indexing all the chunks, if it's >= 0
+	 * then we're only indexing some of the chunks, and leaving the root index
+	 * marked as invalid
+	 */
+	if (info->extended_options.max_chunks > 0)
+		info->extended_options.max_chunks -= 1;
 
-	if (NULL != ht)
+	if (OidIsValid(info->extended_options.barrier_table))
 	{
-		/* Make sure this index is allowed */
-		if (stmt->concurrent)
-			ereport(ERROR,
-					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-					 errmsg("hypertables do not support concurrent "
-							"index creation")));
+		/*
+		 * For insolation tests, and debugging, it's useful to be able to
+		 * pause CREATE INDEX immediately before it starts working on chunks.
+		 * We acquire and immediately release a lock on a barrier table to do
+		 * this.
+		 */
+		Relation barrier = relation_open(info->extended_options.barrier_table, AccessExclusiveLock);
 
-		ts_indexing_verify_index(ht->space, stmt);
+		relation_close(barrier, AccessExclusiveLock);
 	}
+#endif
 
-	ts_cache_release(hcache);
+	/*
+	 * Change user since chunks are typically located in an internal schema
+	 * and chunk indexes require metadata changes. In the single-tranaction
+	 * case, we do this once for the entire table.
+	 */
+	ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+
+	/*
+	 * Hold a lock on the hypertable, it's index, and the chunk to prevent
+	 * from being altered. Since we use the same relids across transactions,
+	 * there is a potential issue if the id gets reassigned between one
```

Comment:
Ah. I think you knew that the lock doesn't do that. anyway, still think we might want to reconsider here. don't really care if there's chunks being added during this operation, even while we're creating an index on a chunk. we hold a lock on the chunk, so any attempt to drop it out from under us will still fail because we have the lock on the chunk.

Reply:
I mean, I could see taking an AccessShare session lock on each chunk at the beginning and then removing the session locks on the chunks at the end of each index creation if you really care that much...

Reply:
That would at least prevent drop_chunks from causing a weird error here...

Reply:
@JLockerman let's talk about this lock monday. Don't think you need the ht lock

Reply:
IIRC, the locks on the hypertable and index are the ones specified by the index creation routine. While I suspect we can eventually relax this, for the first release I’d be happier being overly conservative. I do like the session lock idea though